File: editPool.m
classdef editPool
   properties
      N_BLOCK
      BLOCK_SIZE
      DEL_VAL
      
      normEdit % cell array, each element is edit queue for del, insertion 
      rowEdit % cell array, each element is edit queue for whole row insertion
      
      editFlag % whether a block is editted
      insertRow % index of the insertion row (within block)
   end
   
   methods
       function obj = editPool(nblock, blocksize, delval)
           obj.N_BLOCK = nblock;
           obj.BLOCK_SIZE = blocksize;
           obj.DEL_VAL = delval;
           
           obj.normEdit = cell(1,nblock);
           obj.editFlag = false(1,nblock); % 0-no update for this block, 1-update for this block
           
           obj.insertRow = delval;
           obj.rowEdit = cell(1,nblock);
           
           for i=1:nblock
               obj.normEdit{i} = editQueue(blocksize,4);
               obj.rowEdit{i} = editQueue(1,blocksize);
           end
           
       end
       
       function obj = clear(obj,bi)
           obj.insertRow = obj.DEL_VAL;
           for i=bi:obj.N_BLOCK
               obj.normEdit{i} = clear(obj.normEdit{i});
               obj.editFlag(i) = false;
               
               obj.rowEdit{i} = clear(obj.rowEdit{i});
           end
       end
       
       function obj = insertEditRep(obj, rowInd, ind, val)
           [bi, ii] = getbi(obj,ind);
           obj.normEdit{bi} = insertRep(obj.normEdit{bi}, rowInd, ii, val);
           obj.editFlag(bi) = true;
       end
       
       function obj = insertRowEdit(obj, rowInd, ind, val)
           [bi, ii] = getbi(obj,ind);
           if obj.insertRow==obj.DEL_VAL
               obj.insertRow = rowInd;
           end
           if ~obj.editFlag(bi)
               obj.editFlag(bi)=true;
           end
           obj.rowEdit{bi} = insert(obj.rowEdit{bi}, 1, ii, val);
       end
       
       function obj = sortEdit(obj,bi)
           for bk=bi:obj.N_BLOCK
              if obj.editFlag(bk)
%                   continue
%               else
                  obj.normEdit{bk} = sort(obj.normEdit{bk}); 
              end
           end
       end
       
       function [bi, ii] = getbi(obj,i)
           % get block index of original index i
            n = obj.BLOCK_SIZE;
            bi = floor(i/n)+1;
            ii = mod(i,n);
            if ii==0
                ii=n;
                bi = bi -1;
            end
       end
   end
end

File: common_base/editpool.py
import numpy as np
from common_base.constants import constants

class editQueue:
    def __init__(self, num, dim):
        self.num = num
        self.dim = dim

        self.index = np.zeros((num,dim),constants.DATA_TYPE)
        self.value = np.zeros((num,dim),constants.DATA_TYPE)
        self.pointer = np.zeros(num,constants.DATA_TYPE)
        
    def clear(self):
        self.pointer[:] = 0
        
    def insert(self, rowind, ind, val):
        tp = self.pointer[rowind]
        self.pointer[rowind] = tp+1
        self.index[rowind,tp] = ind
        self.value[rowind,tp] = val
        
    def insert_rep(self,rowind,ind,val):
        tp = self.pointer[rowind]
        for i in range(tp):
            if self.index[rowind,i]==ind:
                self.value[rowind,i] = val;
                return
        self.insert(rowind,ind,val)
       
    def sort(self):
        for ri in range(self.num):
            tp=self.pointer[ri]
            if tp!=0:
                tmpidx = np.argsort(self.index[ri,:tp])
                self.index[ri,:tp] = self.index[ri,tmpidx]
                self.value[ri,:tp] = self.value[ri,tmpidx]

class editPool:
    def __init__(self):
        self.normEdit = [editQueue(constants.BLOCK_SIZE,4) for i in range(constants.N_BLOCK)]
        self.editFlag = np.zeros(constants.N_BLOCK,dtype=bool)
        self.insertRowInd = constants.DEL_VAL;
        self.rowEdit = [editQueue(1,constants.BLOCK_SIZE) for i in range(constants.N_BLOCK)]
        
    def clear(self,bi):
        self.insertRowInd = constants.DEL_VAL
        for i in range(bi,constants.N_BLOCK):
            self.normEdit[i].clear()
            self.editFlag[i] = False
            self.rowEdit[i].clear()
    
    def insert_edit_rep(self, rowind, ind, val):
        bi,ii = constants.getbi(ind)
        self.normEdit[bi].insert_rep(rowind,ii,val)
        self.editFlag[bi] = True
    
    def insert_row_edit(self,rowind,ind,val):
        bi,ii = constants.getbi(ind)
        if self.insertRowInd==constants.DEL_VAL:   
            self.insertRowInd = rowind
        if not self.editFlag[bi]:
            self.editFlag[bi]=True
        self.rowEdit[bi].insert(0,ii,val)
        
    def sort_edit(self,bi):    
        for bk in range(bi,constants.N_BLOCK):
            if self.editFlag[bk]:
                self.normEdit[bk].sort()    
